File: src/slack_stealth_mcp/tools/mark_read.py
"""Tool to explicitly mark conversations as read."""

from __future__ import annotations

from typing import Any

from ..slack_client import SlackAPIError
from ..workspace import WorkspaceManager


async def mark_read(
    manager: WorkspaceManager,
    channel: str,
    timestamp: str | None = None,
    workspace: str | None = None,
) -> dict[str, Any]:
    """Mark a conversation as read.

    This is the ONLY operation that affects read state. Use deliberately
    when you want to mark messages as read.

    Args:
        manager: Workspace manager
        channel: Channel ID to mark as read
        timestamp: Mark as read up to this message timestamp.
                   If not provided, marks all messages as read.
        workspace: Workspace name (uses default if not specified)

    Returns:
        Dictionary with result status
    """
    client = manager.get_client(workspace)

    # If no timestamp provided, get the latest message
    if not timestamp:
        try:
            messages = await client.get_conversation_history(
                channel=channel,
                limit=1,
            )
            if messages:
                timestamp = messages[0].ts
            else:
                return {
                    "ok": True,
                    "channel_id": channel,
                    "message": "No messages to mark as read",
                }
        except SlackAPIError as e:
            return {
                "ok": False,
                "error": f"Failed to get latest message: {e.error}",
                "channel_id": channel,
            }

    # Mark as read
    try:
        await client.mark_conversation(channel=channel, ts=timestamp)

        # Get channel info for context
        channel_name = channel
        try:
            conv = await client.get_conversation_info(channel)
            if conv.is_im and conv.user:
                user = await client.get_user_info(conv.user)
                channel_name = f"DM with @{user.display}"
            else:
                channel_name = conv.display_name
        except Exception:
            pass

        return {
            "ok": True,
            "channel_id": channel,
            "channel_name": channel_name,
            "marked_read_until": timestamp,
            "message": f"Marked {channel_name} as read",
        }

    except SlackAPIError as e:
        return {
            "ok": False,
            "error": f"Failed to mark as read: {e.error}",
            "channel_id": channel,
        }

File: packages/cloudflare/src/tools/mark-read.ts
/**
 * Mark Read Tool - Explicitly mark conversations as read
 *
 * This is the ONLY tool that affects read state.
 */

import { SlackClient, SlackAPIError } from "../slack-client";

interface MarkReadOptions {
  channel: string;
  timestamp?: string;
}

interface MarkReadResult {
  ok: boolean;
  channel_id: string;
  channel_name?: string;
  marked_read_until?: string;
  message?: string;
  error?: string;
}

/**
 * Mark a conversation as read.
 * This is the ONLY operation that affects read state.
 */
export async function markRead(
  client: SlackClient,
  options: MarkReadOptions
): Promise<MarkReadResult> {
  const { channel } = options;
  let { timestamp } = options;

  // If no timestamp provided, get the latest message
  if (!timestamp) {
    try {
      const messages = await client.getConversationHistory(channel, {
        limit: 1,
      });
      if (messages.length > 0) {
        timestamp = messages[0].ts;
      } else {
        return {
          ok: true,
          channel_id: channel,
          message: "No messages to mark as read",
        };
      }
    } catch (error) {
      const apiError = error as SlackAPIError;
      return {
        ok: false,
        channel_id: channel,
        error: `Failed to get latest message: ${apiError.error || apiError.message}`,
      };
    }
  }

  // Mark as read
  try {
    await client.markConversation(channel, timestamp);

    // Get channel info for context
    let channelName = channel;
    try {
      const conv = await client.getConversationInfo(channel);
      if (conv.is_im && conv.user) {
        const user = await client.getUserInfo(conv.user);
        const displayName = user.display_name || user.real_name || user.name;
        channelName = `DM with @${displayName}`;
      } else {
        channelName = conv.name || channel;
      }
    } catch {
      // Ignore - use channel ID as fallback
    }

    return {
      ok: true,
      channel_id: channel,
      channel_name: channelName,
      marked_read_until: timestamp,
      message: `Marked ${channelName} as read`,
    };
  } catch (error) {
    const apiError = error as SlackAPIError;
    return {
      ok: false,
      channel_id: channel,
      error: `Failed to mark as read: ${apiError.error || apiError.message}`,
    };
  }
}
